sharedbuffers/mapped_struct.py

```diff
@@ -342,6 +344,167 @@ def unpack_from(cls, buf, offs, idmap = None):
+    """
+    Gimmick object. Its whole purpose is to implement
+    the protocol expected by the IdMapper classes.
+    """
+
+    def __init__(self, buf, offs):
+        self.buf = buf
+        self.offs = offs
+        self.pos = 0
+
+    def write(self, value):
+        size = len(value)
+        offs = self.offs + self.pos
+        self.buf[offs:offs + size] = value
+        self.pos += size
+        return size
+
+    def tell(self):
+        return self.pos
+
+    def flush(self):
+        pass
+
+    def seek(self, pos):
+        self.pos = pos
+
+
+@cython.cclass
+class proxied_dict(object):
+
+    cython.declare(objmapper=object, vlist=proxied_list)
+
+    def __init__(self, objmapper, vlist):
+        self.objmapper = objmapper
+        self.vlist = vlist
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        ipos = offs
+        offs += 8   # Reserve space for value-list offset.
+        iobuf = BufferIO(buf, offs)
+        mapper_size = ObjectIdMapper.build(_enum_keys(obj), iobuf, map_file=False)
+        offs += mapper_size
+        struct.pack_into("=Q", buf, ipos, mapper_size)
+        return offs + proxied_list.pack_into(obj.values(), buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None):
+        mapper_size, = struct.unpack_from("=Q", buf, offs)
+        offs += 8
+        objmapper = ObjectIdMapper.map_buffer(buf, offs)
+        offs += mapper_size
+        vlist = proxied_list.unpack_from(buf, offs, idmap)
+        return proxied_dict(objmapper, vlist)
+
+    def get(self, key, default_val = None):
+        idx = self.objmapper.get(key)
+        return self.vlist[idx] if idx is not None else default_val
+
+    def __getitem__(self, key):
+        idx = self.objmapper.get(key)
+        if idx is None:
+            raise KeyError(key)
+        return self.vlist[idx]
+
+    def __contains__(self, key):
+        return self.objmapper.get(key) is not None
+
+    def has_key(self, key):
+        return key in self
+
+    def iteritems(self):
+        for key, idx in self.objmapper.iteritems():
+            yield key, self.vlist[idx]
+
+    def items(self):
+        return list(self.iteritems())
+
+    def iterkeys(self):
+        return iter(self.objmapper)
+
+    def keys(self):
+        return list(self.iterkeys())
+
+    def itervalues(self):
+        return iter(self.vlist)
+
+    def values(self):
+        return list(self.itervalues())
+
+    def __iter__(self):
+        return self.iterkeys()
+
+    def __len__(self):
+        return len(self.vlist)
+
+    def _is_eq(self, other):
+        if len(self) != len(other):
+            return False
+
+        for key, val in self.iteritems():
+            if key not in other or other[key] != val:
+                return False
+
+        return True
+
+    def __richcmp__(self, other, op):
+        if op != Py_EQ and op != Py_NE:
+            return False   # No comparison possible
```

Comment:
Python 2 always returns False for operations without meaning (e.g: {} < {}), whereas Python3 raises a TypeError. I'm not sure what's the best option, but I think that since we're aiming to replicate the original dict behavior, going with Python2's behavior is saner (for now).

Reply:
Why don't you just implement `__eq__` and `__ne__` then?

Reply:
> Why don't you just implement `__eq__` and `__ne__` then?

Cython barfs at you if you use those instead of __richcmp__

Reply:
Which version of Cython are you using?

Reply:
> Which version of Cython are you using?

0.26

Reply:
0.27 supports it, maybe we should consider requiring 0.27.0 (0.27.1 has bugs that make it not build)

Otherwise, you can leave __richcmp__ unconditional *and* implement eq and ne, using the import-time monkeypatching trick you committed earlier.

Reply:
BTW, what python does isn't to return False all the time. It compres the `id` of objects instead.